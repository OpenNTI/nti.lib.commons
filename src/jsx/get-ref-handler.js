import Logger from 'nti-util-logger';
const logger = Logger.get('lib:jsx:ref-handler');


export default function getRefHandler (parentRef, localRef) {
	const fn = getRefHandler;
	const REF_HANDLERS = fn.map || (fn.map = new WeakMap());

	if (typeof parentRef !== 'function') {
		if (parentRef) {
			logger.error(`Non-function ref value (${parentRef}) has been dropped. Use a callback instead.`);
		}
		return localRef;
	}

	let h = REF_HANDLERS.get(localRef);
	if (!h) {
		h = (x) => {parentRef(x); localRef(x);};
		REF_HANDLERS.set(localRef, h);
	}
	return h;
}
