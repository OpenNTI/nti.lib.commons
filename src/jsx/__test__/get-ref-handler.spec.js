/* eslint-disable no-console */
/* eslint-env jest */
import { getRefHandler } from '../get-ref-handler.js';

test('getRefHandler', () => {
	const parentRef = jest.fn();
	const localRef = jest.fn();
	jest.spyOn(console, 'error').mockImplementation(() => {});

	expect(() => getRefHandler()).not.toThrow();
	expect(() => getRefHandler(parentRef)).toThrow();
	expect(getRefHandler()).toEqual(void 0);

	expect(getRefHandler({}, localRef)).toBe(localRef);
	expect(console.error.mock.calls).toMatchSnapshot();

	const A = getRefHandler(parentRef, localRef);
	const B = getRefHandler(parentRef, localRef);
	const C = getRefHandler(jest.fn(), localRef);
	const D = getRefHandler(parentRef, jest.fn());

	expect(A).toBe(B);

	expect(C).not.toBe(A);
	expect(D).not.toBe(B);
	expect(D).not.toBe(C);
});
