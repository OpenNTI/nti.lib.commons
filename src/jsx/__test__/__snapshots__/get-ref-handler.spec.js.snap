// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRefHandler 1`] = `
Array [
  Array [
    "Unsupported value (object: [object Object]) has been dropped. Use a ref object or callback instead.",
  ],
]
`;
